pkg/discovery/processor/volume_processor.go

```diff
@@ -0,0 +1,115 @@
+package processor
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/turbonomic/kubeturbo/pkg/cluster"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/repository"
+	v1 "k8s.io/api/core/v1"
+)
+
+type VolumeProcessor struct {
+	ClusterInfoScraper cluster.ClusterScraperInterface
+	KubeCluster        *repository.KubeCluster
+}
+
+func NewVolumeProcessor(kubeClient cluster.ClusterScraperInterface,
+	kubeCluster *repository.KubeCluster) *VolumeProcessor {
+	return &VolumeProcessor{
+		ClusterInfoScraper: kubeClient,
+		KubeCluster:        kubeCluster,
+	}
+}
+
+func (p *VolumeProcessor) ProcessVolumes() {
+	clusterName := p.KubeCluster.Name
+	pvList, err := p.ClusterInfoScraper.GetAllPVs()
+	if err != nil {
+		glog.Errorf("Failed to get persistent volumes for cluster %s: %v.", clusterName, err)
+		return
+	}
+	glog.V(2).Infof("There are %d volumes.", len(pvList))
+
+	pvcList, err := p.ClusterInfoScraper.GetAllPVCs()
+	if err != nil {
+		glog.Errorf("Failed to get persistent volumes claims for cluster %s: %v.", clusterName, err)
+		return
+	}
+	glog.V(2).Infof("There are %d volume claims.", len(pvcList))
+
+	allPods, err := p.ClusterInfoScraper.GetAllPods()
+	if err != nil {
+		glog.Errorf("Failed to get all pods for cluster %s: %v.", clusterName, err)
+		return
+	}
+
+	// A map which lists the volume to pvc bindings (1 to 1)
+	// There can be volumes which are not bound via claims
+	// and pods use them directly.
+	volumeToClaimMap := make(map[*v1.PersistentVolume]*v1.PersistentVolumeClaim)
+	for _, pv := range pvList {
+		found := false
+		for _, pvc := range pvcList {
+			if pvc.Spec.VolumeName == pv.Name {
+				volumeToClaimMap[pv] = pvc
+				found = true
+				break
+			}
+		}
+		if !found {
+			volumeToClaimMap[pv] = nil
+		}
+	}
+
+	// TODO (irfanurrehman): We might need the volume name to map the metrics generated from pod
+	// and gathered via kubelet monitor to the individual volumes.
+	// TODO (irfanurrehman): (Next Step) To find out how are metrics generated for volumes which
+	// are used by multiple pods.
+	volumeToPodsMap := make(map[*v1.PersistentVolume][]repository.PodVolume)
+	for pv, pvc := range volumeToClaimMap {
+		if pvc == nil {
+			// Unused volume.
+			volumeToPodsMap[pv] = nil
+		}
```

Comment:
Do you want to add a "continue" here to skip the whole iteration below when pvc == nil? And then in line 79 no need to check `if pvc != nil`

Reply:
This is intentionally left this way to remember and handle the TODO in the else case in the loop below.
`// TODO (irfanurrehman): check other volume sources categorically.`

Reply:
I see. Make sense